thorlcr/slave/traceslave.hpp

```diff
@@ -0,0 +1,203 @@
+                            DBGLOG("DEBUG: Exception: Row size %u", (unsigned) recSize);
+                    }
+                    xmlwrite.outputEndNested("Row", true);
+                    mb.append(xmlwrite.str());
+                    xmlwrite.clear();
+                }
+                if (++rowPos==rowBufferSize) rowPos = 0;
+            }
+        }
+    };
+    const unsigned traceQueueSize;
+    class TraceQueue buffers[2];
+    class TraceQueue rowBufferLogCache;
+    atomic_t rowBufInUse;
+    Owned<IThorDataLink> thorDataLink;
+    IHThorArg *helper;
+    atomic_t lastSequence;
+
+    inline void enqueueRowForTrace(const void *row)
+    {
+        buffers[atomic_read(&rowBufInUse)].enqueue(row,atomic_add_and_read(&lastSequence,1));
+    }
+
+public:
+    virtual void start()
+    {
+        thorDataLink->start();
+    }
+    virtual bool isGrouped() { return thorDataLink->isGrouped(); }
+    virtual const void *nextRowGE(const void * seek, unsigned numFields, bool &wasCompleteMatch, const SmartStepExtra &stepExtra)
+    {
+        const void *row = thorDataLink->nextRowGE(seek, numFields, wasCompleteMatch, stepExtra);
+        enqueueRowForTrace(row);
+        return row;
+    }    // can only be called on stepping fields.
+    virtual IInputSteppingMeta *querySteppingMeta() { return thorDataLink->querySteppingMeta(); }
+    virtual bool gatherConjunctions(ISteppedConjunctionCollector & collector) { return thorDataLink->gatherConjunctions(collector); }
+    virtual void resetEOF() { thorDataLink->resetEOF(); }
+
+    virtual void getMetaInfo(ThorDataLinkMetaInfo &info) { thorDataLink->getMetaInfo(info); }
+    virtual CActivityBase *queryFromActivity() {return thorDataLink->queryFromActivity(); } // activity that has this as an output
+    virtual void dataLinkSerialize(MemoryBuffer &mb) {thorDataLink->dataLinkSerialize(mb); }
+    virtual unsigned __int64 queryTotalCycles() const { return thorDataLink->queryTotalCycles(); }
+
+    virtual const void *nextRow()
+    {
+        const void *row = thorDataLink->nextRow();
+        enqueueRowForTrace(row);
+        return row;
+    }
+    virtual void stop()
+    {
+        thorDataLink->stop();
+    }
+
+    inline const void *ungroupedNextRow() { return thorDataLink->ungroupedNextRow(); }
+
+    virtual void setInputSplitterOutputBase(CSplitterOutputBase *_input, unsigned _id=0)
+    {
+        thorDataLink->setInputSplitterOutputBase(_input, _id);
+    }
+    virtual void debugRequest(MemoryBuffer &mb)
+    {
+        // NOTE - cannot be called by more than one thread
+        buffers[1].init(traceQueueSize+1);
+        rowBufferLogCache.init(traceQueueSize);
+        int bufToDump = atomic_read(&rowBufInUse);
+
+        int inactiveBuf = 1 - bufToDump;
+        buffers[inactiveBuf].queueOut(rowBufferLogCache);
+
+        atomic_set(&rowBufInUse, inactiveBuf);// Swap Active & inactiveBuf
+        buffers[bufToDump].queueOut(rowBufferLogCache);
```

Comment:
I think the 2 queueOut calls follow from our previous conversation:
```
Also (and maybe we discussed this(?), it does mean that if there's that race condition, that rows will appear out of sequence. e.g.:

buffers[0] = [a],[b],*c
dump swaps, a,b added to tracing, buffers[1] now active
buffers[1] = [d],[e],*f
dump swap, d,e added to tracing, buffers[0] now active
buffers[0] = [],[],[c],[g],[h],*i
dump swap, c,g,h added to tracing, buffers[1] now active
buffers[1] = [],[],[f],[j],[k],*l
dump swap, f,j,k added to tracing..
Should we just accept that, and just put a comment to mention that this can happen?

Or could we look at the other buffer (the one about to be swapped in) for the trailing row and add to the tracing circular ring, before swapping in ?
I think that could mean there's still a window if 2 swaps happened very quickly, where it would still be active and not picked up and therefore still lead to out order rows, but in practice with the time taken between 2 swaps over 2 debugRequest() hops that isn't going to happen.
```

It would be worth adding a comment to say why callling queueOut for existing (swapped out) buffer, then again for newly swapped out buffer.